modules/core/src/effects/lighting-effect.js

```diff
@@ -68,18 +70,28 @@ export default class LightingEffect extends Effect {
     const viewport = layer.context.viewport;
     const {coordinateSystem, coordinateOrigin} = layer.props;
     const projectedPointLights = [];
+    let position;
 
     for (let i = 0; i < this.pointLights.length; i++) {
       const pointLight = this.pointLights[i];
-      const position = projectPosition(pointLight.position, {
-        viewport,
-        coordinateSystem,
-        coordinateOrigin,
-        fromCoordinateSystem: viewport.isGeospatial
-          ? COORDINATE_SYSTEM.LNGLAT
-          : COORDINATE_SYSTEM.IDENTITY,
-        fromCoordinateOrigin: [0, 0, 0]
-      });
+      if (pointLight instanceof CameraLight) {
```

Comment:
It's a little concerning that the effect class has to know about `CameraLight` for it to work. Can users not implement their own light sources?

We could defer to each light source class to calculate this, e.g.

```js
pointLight.getPosition(parameters);
```

Reply:
Lighting effect is owner of different light sources, I think it is reasonable for it to know each kind of light source,  default behaviors and doing some jobs based on that.

User can implement their own light sources, but we need define the light source category and the basic behavior first. For example, user can implement their own CamerLight by extending from our CameraLight source, we will provide the internal camera position automatically as a predefined behavior, and user don't need understand our complicated projection system.

Point light sources stores the pre-projection position, color, and intensity ,etc. If we put this project logic inside light sources, there will be a coupling between viewport, projection system and the light source classes, which is not a good design from my perspective. Also the projected position is not the same under different viewport & zoom, then the getPosition API is even more ambiguous with those complicated logics lives together.

Reply:
Going to move viewport related logic to light source classes per offline discussion in case user need implement their own light source category.